storage/paths.go

```diff
@@ -75,28 +123,99 @@ func (pm *pathMapper) path(spec pathSpec) (string, error) {
 	repoPrefix := append(rootPrefix, "repositories")
 
 	switch v := spec.(type) {
-	case manifestTagsPath:
-		return path.Join(append(repoPrefix, v.name, "manifests")...), nil
-	case manifestPathSpec:
-		// TODO(sday): May need to store manifest by architecture.
-		return path.Join(append(repoPrefix, v.name, "manifests", v.tag)...), nil
-	case layerLinkPathSpec:
-		components, err := digestPathComoponents(v.digest)
+
+	case manifestRevisionPathSpec:
+		components, err := digestPathComponents(v.revision, false)
 		if err != nil {
 			return "", err
 		}
 
-		// For now, only map tarsum paths.
-		if components[0] != "tarsum" {
-			// Only tarsum is supported, for now
-			return "", fmt.Errorf("unsupported content digest: %v", v.digest)
+		return path.Join(append(append(repoPrefix, v.name, "manifests", "revisions"), components...)...), nil
```

Comment:
Expanding on this: most of the directory components actually create their parent and call the mapper, avoiding repetition in the path component names. I think the only repeat here is for the tags and manifest revisions directory.